Merge snapshot module and create_layout module into one

SolidifiedRay opened the pull request:
In PR #46, @IshaDave and I created a new module called [`before_after_filesystem_snapshot.py`](https://github.com/in-toto/layout-web-tool/pull/46/files#diff-878f78afb6c64a06f28f3ec6474964e6). This module returns a 'snapshot' of a file structure showing which files were unchanged, modified, added, or removed. It also includes a function that generates material rules and product rules based on this snapshot.

In this PR, I move the functions from `before_after_filesystem_snapshot.py` to `create_layout.py`. The `create_layout.py` will now utilize the snapshot function to generate more complex material rules and product rules, including `MATCH`, `ALLOW`, `MODIFY`, `DELETE`, `CREATE`, `DISALLOW`.

You can also read issue #48 for more ideas.

adityasaky commented on create_layout.py:
```diff
@@ -91,34 +91,83 @@
 
-def create_material_rules(links, index):
+def snapshot(before_dict, after_dict):
+  '''before_after_snapshot is a simple function that returns which files were
```
```suggestion
  '''A simple function that returns which files were
```

adityasaky commented on create_layout.py:
```diff
@@ -91,34 +91,83 @@
+
+  # Returning the snapshot of the new file system
+  return (sorted(unchanged_files), sorted(modified_files), sorted(added_files),
+  sorted(removed_files))
```
```suggestion
      sorted(removed_files))
```

adityasaky commented on create_layout.py:
```diff
@@ -91,34 +91,83 @@
+  return (sorted(unchanged_files), sorted(modified_files), sorted(added_files),
+  sorted(removed_files))
+
+def create_material_rules(current_snapshot, links, index):
```
Let's add what each argument means in the docstring too.

adityasaky commented on create_layout.py:
```diff
@@ -91,34 +91,83 @@
+    pre_step_name = links[index-1].name
+    pre_snapshot = snapshot(links[index-1].materials, links[index-1].products)
+
+    for i in range(3):
```
Any reason for `3` being hard coded? :)

adityasaky replied:
Instead, let's attach each return of `snapshot` to a separate variable?

```
pre_unchanged_files, pre_modified_files, pre_added_files, pre_removed_files = snapshot(...)
```

what do you think?

lukpueh replied:
+1 for tuple unpacking!

adityasaky commented:
cc @lukpueh could you take a look at this? :smile:

lukpueh reviewed (commented):
Thanks for working on this, @SolidifiedRay! It would be really great if your commit history made it clearer, what parts of the PR are renames (e.g. `test_before_after_filesystem_snapshot.py` --> `test_create_layout.py` ), or copy-pasting entire functions from one module to another (e.g. the `snapshot` function), and what changes/adds functionality (e.g. the `create_layout.create_*_rules` functions work differently now and there also seem to be new tests). Do you think you can redraw the git history? 

Our [commits guideline document](https://github.com/secure-systems-lab/lab-guidelines/blob/master/commits.md) has a corresponding paragraph. It also has some good recommendations for commit messages. ;)

SolidifiedRay commented:
> Do you think you can redraw the git history?

Yes, I will do this. Thanks for pointing this out!

SolidifiedRay committed:
Move snapshot function from before_after_filesystem_snapshot.py to create_layout.py

Move snapshot function from before_after_filesystem_snapshot.py to create_laytou.py and remove before_after_filesystem_snapshot.py.
This snapshot function will be used to enhance the rules generation function in create_layout.py.

SolidifiedRay committed:
Enhance rule generation function

- Rename snapshot() function to changes_between_snapshots()
- Generate DELETE, MODIFY, and CREATE rules generation
- For MATCH rules, match only those that already were in the previous step, and allow the rest by name

SolidifiedRay committed:
Rename test_before_after_filesystem_snapshot.py to test_create_layout.py

SolidifiedRay committed:
Add tests for rule generation function

- Add test_create_material_rules_with_zero_index
- Add test_create_material_rules_with_nonzero_index
- Add test_create_product_rules

adityasaky commented on create_layout.py:
```diff
@@ -91,45 +91,120 @@
 
   expected_materials = []
+  unchanged_files, modified_files, _, deleted_files = changes_of_curr_step
+  curr_step_materials = unchanged_files + modified_files + deleted_files
```
I may be missing something here, but `DELETE` doesn't apply as a material rule, but rather as a product rule for artifacts that are deleted as part of that corresponding step.

```DELETE: indicates that materials matched by the pattern must not appear as products of this step.```

SolidifiedRay replied:
I discussed this with Lukas in the previous PR, and he suggests me to put `DELETE` in the materials rules.
https://github.com/in-toto/layout-web-tool/pull/46#discussion_r464405264

I think what he means is that becuase deleted artifacts won't show up in the product queue, we should put `DELET` in the materilas rules. What do you think?

adityasaky replied:
Yeah, looks like I'd misunderstood how the delete rule applies in practice. Carry on!

adityasaky commented on create_layout.py:
```diff
@@ -91,45 +91,120 @@
+  return (sorted(unchanged_files), sorted(modified_files), sorted(added_files),
+      sorted(removed_files))
+
+def create_material_rules(changes_of_curr_step, links, index):
```
I wonder if it'd be saner to define the `create_material_rules` API to be unaware of the index. It should accept two successive links and return the material rules, no? It can even call `changes_between_snapshots` internally using the links passed. If a link is the first in the supply chain, the "preceding" link can be `None`.

```
def create_material_rules(previous_link, current_link):
  ....
```

Does that make sense? I feel like it would simplify this API since this would be called by some iterative block of code anyway, in this instance `create_layout_from_ordered_links`.

cc @lukpueh

adityasaky commented on create_layout.py:
```diff
@@ -91,45 +91,120 @@
 
-def create_product_rules(links, index):
-  """Create generic product rules (2 variants)
+def create_product_rules(changes_of_curr_step):
```
If we follow the suggestion for `create_material_rules`, we can probably mirror the signature here and call the snapshot function internally.

adityasaky reviewed (changes requested):
This is looking pretty good, @SolidifiedRay! Just some thoughts on how we can simplify the APIs further. :smile:

adityasaky commented on create_layout.py:
```diff
@@ -141,10 +138,11 @@ def create_material_rules(changes_of_curr_step, links, index):
   curr_step_materials = unchanged_files + modified_files + deleted_files
 
-  if index == 0:
+  if previous_link == None:
```
```suggestion
  if previous_link is None:
```

adityasaky commented on create_layout.py:
```diff
@@ -91,45 +91,116 @@
+    # MATCH only those that already were in the previous step
+    # Allow the rest by name
+    pre_step_name = previous_link.name
+    pre_unchanged_files, pre_modified_files, pre_added_files, _ = \
```
I may be missing something, but why is this necessary? Can't we just use `previous_link.products`?

SolidifiedRay replied:
Yes, you are right. Thanks for pointing this out!

adityasaky commented on .travis.yml:
```diff
@@ -5,4 +5,4 @@ cache: pip
   - pip install -r requirements.txt
 script:
-  python -m unittest tests.test_before_after_filesystem_snapshot
+  python -m unittest tests.test_create_layout
```
cc @IshaDave are we moving to using tox?

IshaVDave replied:
Hi. So maybe you already know by know, but @lukpueh and I had a conversation earlier and no, we're not moving to tox for the layout-web-tool.

adityasaky commented on create_layout.py:
```diff
@@ -91,45 +91,111 @@
 
-def create_material_rules(links, index):
+def changes_between_snapshots(before_dict, after_dict):
+  '''Given two 'snapshots' of a file structure -- 'before' and 'after' -- 
```
I think it's worth being ambiguous about whether they're files or not. Here, we don't really care about the type of the artifacts. Let's just go with artifacts throughout?

adityasaky commented on create_layout.py:
```diff
@@ -91,45 +91,111 @@
+def create_material_rules(previous_link, current_link):
   """Create generic material rules (3 variants)
 
+  previous_link: the link of previous step
```
I understand we're using the google docstring format as @lukpueh suggests in https://github.com/secure-systems-lab/code-style-guidelines/issues/20.

adityasaky commented on create_layout.py:
```diff
@@ -91,45 +91,111 @@
+    Both these dictionaries have file
+    names as the keys and their hashes as the values.'''
+
+  unchanged_files = set()
```
So, I think I understand why this was updated to use sets, but I think it's worth noting that `sorted` below converts it back into a list. So I only see a reason to use sets if we want to get rid of the multiple `for` loops.

adityasaky replied:
I do want to emphasize that you don't have to do it this way, but what I meant when I mentioned sets here is something along the lines of:

```
before_set = set(before_dict.keys())
after_set = set(after_dict.keys())

removed_keys = sorted(before_set.difference(after_set))
added_keys = sorted(after_set.difference(before_set))

for key in before_set.intersection(after_set)):
  if before_dict[key] == after_dict[key]:
    unchanged_files.append(key) # sort later
  else:
    modified_files.append(key) # sort later
```

SolidifiedRay replied:
If I remember correctly, the only reason we return sorted lists here is for the unittest. If we don't sort the list, `changes_between_snapshots` will return lists in random order. As a result, `assertEqual` will return false even when two lists are considered to be "equal".
Isha wrote this as a quick fix to the problem above, but I don't think it is necessary to sort the list here. Maybe we can just return sets instead, and use `assertSetEqual`. Or maybe we can still return lists, but we sort them in the unittest function. What do you think?

adityasaky replied:
Sorting here or elsewhere is fine when we're dealing with tests. Alternatively, if we go full sets, `assertSetEqual` is fine. I just meant to say that this particular snippet isn't making use of the features of a set over a list.

```
before_set = set(before_dict.keys())
after_set = set(after_dict.keys())

removed_keys = before_set.difference(after_set)
added_keys = after_set.difference(before_set)

for key in before_set.intersection(after_set)):
  if before_dict[key] == after_dict[key]:
    unchanged_files.add(key)
  else:
    modified_files.add(key)
```

Perhaps we can do something like this and use assertSetEqual in the test?

SolidifiedRay replied:
Thanks again for the feedback @adityasaky ! I have force pushed the branch. Please let me know if you feel something is not right.

adityasaky commented on create_layout.py:
```diff
@@ -91,45 +91,111 @@
+  previous_link: the link of previous step
+  current link: the link of current step
+
   * MATCH available materials with products from previous step (links must be an
```
we're no longer passing a list of links

adityasaky commented on create_layout.py:
```diff
@@ -91,45 +91,111 @@
+      expected_materials.append(["ALLOW", file])
+    for file in deleted_files:
+      expected_materials.append(["DELETE", file])
     expected_materials.append(["DISALLOW", "*"])
```
This shouldn't be within the condition, I think, and it needs to go at the end?

adityasaky commented on create_layout.py:
```diff
@@ -91,45 +91,111 @@
+    for file in curr_step_materials.difference(previous_link.products):
+        expected_materials.append(["ALLOW", file])
+
+    for file in deleted_files:
```
This is not dependent on the condition, it can live independently outside it instead of twice, in each branch.

SolidifiedRay replied:
I may be missing something. Do you mean when we `MATCH` an artifact, we should also `ALLOW` it? I put `ALLOW` in different branch because I think we don't need to `ALLOW` an artifact if we already `MATCH` it with the artifact from the previous step.

adityasaky replied:
No I meant the DELETE rule.

adityasaky commented on tests/test_create_layout.py:
```diff
@@ -0,0 +1,183 @@
+import create_layout
+import in_toto.models.link
+
+class Test_before_after_filesystem_snapshot(unittest.TestCase):
```
Test rename necessary

adityasaky commented on tests/test_create_layout.py:
```diff
@@ -0,0 +1,183 @@
+      ['MATCH', 'one.tgz', 'WITH', 'PRODUCTS', 'FROM', 'first_step'],
+      ['MATCH', 'foo/two.tgz', 'WITH', 'PRODUCTS', 'FROM', 'first_step'],
+      ['MATCH', 'three.txt', 'WITH', 'PRODUCTS', 'FROM', 'first_step'],
+      ['DELETE', 'three.txt']
```
Missing a DISALLOW here, possibly because of the conditional problem I highlighted above. If not, there may be another place to add a DISALLOW! :)

adityasaky reviewed (changes requested):
Solid work, @SolidifiedRay! This is looking pretty good. I quite like the tests you have, and the PR needs a few mostly straightforward changes IMO.

SolidifiedRay commented:
Thanks a lot for the review @adityasaky ! I have made most of the changes but I still have some questions and I have replied to your review.

adityasaky reviewed (approved):
LGTM, thanks @SolidifiedRay!

cc @lukpueh any thoughts?

adityasaky commented on create_layout.py:
```diff
@@ -88,48 +88,118 @@
   expected_materials = []
+  unchanged_artifacts, modified_artifacts, _, deleted_artifacts = \
+      changes_between_snapshots(current_link.materials, current_link.products)
+  curr_step_materials = set(itertools.chain(unchanged_artifacts,
```
nit: can't this use `.union` or `.extend` (if list)?

SolidifiedRay replied:
Yes. `.union` works. Thanks for pointing this out!

lukpueh commented on create_layout.py:
```diff
@@ -91,45 +91,114 @@
+
+  Args:
+    previous_link: a link of previous step,
+        including previous step's materials and prodcuts
```
```suggestion
        including previous step's materials and products
```

lukpueh commented on create_layout.py:
```diff
@@ -91,45 +91,114 @@
+    previous_link: a link of previous step,
+        including previous step's materials and prodcuts
+    current link: a link of current step,
+        including current step's materials and prodcuts
```
```suggestion
        including current step's materials and products
```

lukpueh commented on create_layout.py:
```diff
@@ -91,45 +91,114 @@
-  expected_products = []
+  Args:
+    curretn_link: a link of current step,
+        including current step's materials and prodcuts
```
```suggestion
        including current step's materials and products
```

lukpueh commented on create_layout.py:
```diff
@@ -91,45 +91,114 @@
+    which have been removed, which have been modified, 
+    and which have remained unchanged. 
+    Both these dictionaries have artifact
+    names as the keys and their hashes as the values.'''
```
I suggest to use `"""` consistently for docstrings, also I don't fully understand the line wrap here and also in some cases below. I'd just wrap docstrings at 79 characters.

lukpueh commented on create_layout.py:
```diff
@@ -91,45 +91,114 @@
-
-  * MATCH available materials with products from previous step (links must be an
+def changes_between_snapshots(before_dict, after_dict):
+  '''Given two 'snapshots' of a artifacts structure -- 'before' and 'after' -- 
```
```suggestion
  '''Given two 'snapshots' of an artifacts structure -- 'before' and 'after' --
```

lukpueh replied:
Note the removal of trailing whitespace in the suggestion above. Can you do this here and elsewhere in your PR? (See `git diff --check`, or your editors trailing-whitespace display setting)

lukpueh commented on create_layout.py:
```diff
@@ -91,45 +91,114 @@
-  * MATCH available materials with products from previous step (links must be an
+def changes_between_snapshots(before_dict, after_dict):
+  '''Given two 'snapshots' of a artifacts structure -- 'before' and 'after' -- 
+    return a tuple specifying which artifact have been added, 
```
```suggestion
    return a tuple specifying which artifacts have been added, 
```

lukpueh commented on create_layout.py:
```diff
@@ -91,45 +91,114 @@
+    else:
+      modified_artifacts.add(key)
+
+  # Returning the snapshot of the new artifacts system
```
I think the diff of two 'snapshots' is not really a snapshot itself. Let's maybe just remove the comment, the code seems clear enough. :)

lukpueh commented on create_layout.py:
```diff
@@ -91,45 +91,114 @@
+    if before_dict[key] == after_dict[key]:
+      unchanged_artifacts.add(key)
+    else:
+      modified_artifacts.add(key)
```
🥇 for switching to set operations.

lukpueh replied:
You could even use set operations to get `unchanged_artifacts` and `modified_artifacts`. But I think the for-loop is more readable. ;)

lukpueh commented on create_layout.py:
```diff
@@ -91,45 +91,114 @@
+    # MATCH only those that already were in the previous step
+    for artifact in curr_step_materials.intersection(previous_link.products):
+      expected_materials.append(
+        ["MATCH", artifact, "WITH", "PRODUCTS", "FROM", previous_link.name])
```
```suggestion
          ["MATCH", artifact, "WITH", "PRODUCTS", "FROM", previous_link.name])
```

lukpueh commented on create_layout.py:
```diff
@@ -91,45 +91,114 @@
-  return expected_materials
+    # Allow the rest by name
+    for artifact in curr_step_materials.difference(previous_link.products):
+        expected_materials.append(["ALLOW", artifact])
```
```suggestion
      expected_materials.append(["ALLOW", artifact])
```

lukpueh commented on create_layout.py:
```diff
@@ -91,45 +91,114 @@
 
-  expected_products = []
+  Args:
+    curretn_link: a link of current step,
```
```suggestion
    current_link: a link of current step,
```

lukpueh commented on tests/test_create_layout.py:
```diff
@@ -0,0 +1,206 @@
+    }
+  }
+
+  empty_set = set()
```
I think it would be okay to just instantiate a new `set()` each time you compare against an empty set. But this is fine too.

lukpueh commented on tests/test_create_layout.py:
```diff
@@ -0,0 +1,206 @@
+      ['ALLOW', 'bar/bat/four.tgz'],
+      ['ALLOW', 'one.tgz'],
+      ['ALLOW', 'foo/two.tgz'],
+      ['ALLOW', 'three.txt'],
```
This ALLOW rule makes the subsequent DELTE rule moot. We should only 'ALLOW' artifacts that aren't 'DELETED' afterwards.

lukpueh commented on tests/test_create_layout.py:
```diff
@@ -0,0 +1,206 @@
+      ['MATCH', 'bar/bat/four.tgz', 'WITH', 'PRODUCTS', 'FROM', 'first_step'],
+      ['MATCH', 'one.tgz', 'WITH', 'PRODUCTS', 'FROM', 'first_step'],
+      ['MATCH', 'foo/two.tgz', 'WITH', 'PRODUCTS', 'FROM', 'first_step'],
+      ['MATCH', 'three.txt', 'WITH', 'PRODUCTS', 'FROM', 'first_step'],
```
Ha! This is rather interesting. Like above, this rule makes the subsequent 'DELETE' rule moot. But unlike above, both rules here are actually meaningful. We should document this issue and add it to the list of artifact rule discussion items in https://github.com/in-toto/docs/issues/4. (cc @SantiagoTorres)

SolidifiedRay replied:
Thanks again for pointing this interesting problem out! 
For this problem, I believe the DELETE rule is more important. I think a malicious artifact won't cause many problems if we are going to delete it. But I also believe it is important to let the user know that something wrong happened in the supply chain if there is a mismatch of the artifact.

lukpueh replied:
So you say, if an artifact gets deleted in one step, it doesn't really matter if it matched the product from a previous step? That's an interesting thought, and I do agree with you on a practical level. 

However, I still think it is a lot more important to guarantee the integrity of artifacts in transit between steps than within a step. Because I trust the functionary more than the man in the middle. So I would say, match rules across steps should always be prioritized. 

Moreover, a `DISALLOW *` in the products rules of a step could catch any artifact that was expected to be deleted but wasn't actually deleted.

@SolidifiedRay, would you mind filing an issue on [in-toto/docs](https://github.com/in-toto/docs/issues) for this, and see if you can find other such conflicting rules too? Let's discuss this with other in-toto folks there before make a decision here. In the meanwhile we can just warn the caller about these cases and suggest to manually review.

SolidifiedRay replied:
Sure! Here is the link: [issue#41](https://github.com/in-toto/docs/issues/41).

lukpueh replied:
Excellent write-up of the issue. Thanks, a lot!

lukpueh commented on tests/test_create_layout.py:
```diff
@@ -0,0 +1,206 @@
+    ]
+
+    # create_material_rules returns a lists with random order,
+    # so we need to sort the lists
```
That is a problem. The order of a list of artifact rules does matter. For instance, if you call `'DISALLOW' *` first, then every subsequent rule is moot. 

I think we should order the artifacts before we iterate over them in `create_material_rules` and `create_product_rules`.

SolidifiedRay replied:
Sorry, I didn't say clear enough here. The rules still follow the correct order, but the order of artifacts inside each group of rules is random. For example, in a material rule, the order of matching different artifacts is random. But this material rule still keeps all the "MATCH" rules the first, and the "DISALLOW *" rule the last.

Or do you mean the order of artifacts inside each group of rules is also important? What I concern about is that sorting the artifacts will take extra time.

lukpueh replied:
Understood. And no, I don't think the order of artifacts inside each group is important. This is more a problem of the tests, which don't verify if `MATCH` rules are first, `DISALLOW *` is last, etc., right? And it is also a problem of determinism. It would be nice to get the same results across different Python versions. 

I wouldn't worry too much about the extra time ordering will take (let's cross the performance optimization bridge when we get there).

SolidifiedRay replied:
I have implemented the sort. Please check it. Thanks! @lukpueh

lukpueh commented on tests/test_create_layout.py:
```diff
@@ -0,0 +1,206 @@
+    ]
+
+    # create_product_rules returns a lists with random order,
+    # so we need to sort the lists
```
See comment about ordering above.

lukpueh reviewed (changes requested):
Excellent work, @SolidifiedRay and 💯  for extensive testing!! At first I thought I'd only point out some minor style nits, but looking at the tests I did find some interesting issues with the rule creation functions (which is good!! :). 

So it seems like we need to be more careful, when an artifact qualifies for multiple rules (see inline comments in tests).

SolidifiedRay commented:
Thanks a ton for the review! @lukpueh  I have fixed most of the nits that you mentioned. The issue related to the rule creation function is interesting, and I left some ideas under your comments. Could you please check on it? Thanks!

lukpueh commented on create_layout.py:
```diff
@@ -91,45 +91,112 @@
-  NOTE: Read header docstring for ideas for more complexity.  """
+def changes_between_snapshots(before_dict, after_dict):
+  """Given two 'snapshots' of an artifacts structure -- 'before' and 'after' --
+  return a tuple specifying which artifacts  have been added, which have been
```
```suggestion
  return a tuple specifying which artifacts have been added, which have been
```

lukpueh commented on create_layout.py:
```diff
@@ -91,45 +91,112 @@
+          ["MATCH", artifact, "WITH", "PRODUCTS", "FROM", previous_link.name])
+
+    # Allow the rest by name
+    for artifact in curr_step_materials.difference(previous_link.products) \
+        .difference(deleted_artifacts):
+        expected_materials.append(["ALLOW", artifact])
```
```suggestion
    for artifact in curr_step_materials.difference(previous_link.products) \
        .difference(deleted_artifacts):
      expected_materials.append(["ALLOW", artifact])
```

lukpueh commented:
Thanks for addressing all my comments, and for sharing your suggestions about conflicting rules, @SolidifiedRay! As said above, I think we don't need to find a solution here. Let's just add a warning about the conflict so that we can move forward here.

lukpueh commented on create_layout.py:
```diff
@@ -91,48 +91,123 @@
+      modified_artifacts.add(key)
+
+  return (sorted(unchanged_artifacts), sorted(modified_artifacts), sorted(added_artifacts),
+      sorted(removed_artifacts))
```
Is it necessary to convert the sets to ordered lists here? Don't you convert them back to sets in the calling functions anyway?

To solve the determinism issue we discussed, I think it would be easier to just do: `for artifact in sorted(...): ...` in all the for-loops in your `create_*_rules` functions

lukpueh commented on create_layout.py:
```diff
@@ -91,48 +91,123 @@
+  expected_delete_rules = []
+  unchanged_artifacts, modified_artifacts, _, deleted_artifacts = \
+      changes_between_snapshots(current_link.materials, current_link.products)
+  curr_step_materials = set(unchanged_artifacts + modified_artifacts \
```
FYI, no need to escape line break inside parantheses.

lukpueh commented on create_layout.py:
```diff
@@ -91,48 +91,123 @@
-  NOTE: Read header docstring for ideas for more complexity.  """
+  else:
+    # MATCH only those that already were in the previous step
+    # WARNING: if we have a MATCH rule and a DELETE rule on the same artifact,
```
Oh, I meant an actual warning using e.g. the `logging` module here, not just a comment. :)

lukpueh commented on create_layout.py:
```diff
@@ -91,48 +91,123 @@
+  expected_materials_rules = []
+  expected_allow_rules = []
+  expected_match_rules = []
+  expected_delete_rules = []
```
What are all these new lists for? I think it's fine to just add the rules to `expected_materials_rules` one by one as long as you iterate over ordered artifacts. Or am I missing something?

SolidifiedRay commented:
I have reimplemented the sort and added a warning about the rule conflict. I really appreciate your code review all the time. Thanks! @lukpueh

lukpueh commented on create_layout.py:
```diff
@@ -88,51 +88,125 @@
+    # CREATE added artifacts
+    expected_products_rules.append(["CREATE", artifact])
+  # DISALLOW everything else
+  expected_products_rules.append(["DISALLOW", "*"])
```
Shouldn't you sort artifacts here too?

lukpueh commented on create_layout.py:
```diff
@@ -88,51 +88,125 @@
-    expected_materials.append(["DISALLOW", "*"])
+    # If MATCH rule and DELETE rule are applied to the same artifact
+    if deleted_artifacts.intersection(previous_link.products):
+      warnings.warn("DELET rule is moot because of the previous MATCH rule."
```
```suggestion
      warnings.warn("DELETE rule is moot because of the previous MATCH rule."
```

lukpueh commented on create_layout.py:
```diff
@@ -88,51 +88,125 @@
+  # DISALLOW all other artifacts
+  expected_materials_rules.append(["DISALLOW", "*"])
 
+  return expected_materials_rules
```
I think you could simplify the control flow in this function a little bit, here is what I would do:

```
- if there was a previous step, add match rules for all materials that were products in the previous step
- add delete rules for all deleted artifacts, 
  but warn for any delete rule that has no effect because of a previous match rule
- add allow rules for all remaining materials
- add disallow *
```

lukpueh commented on create_layout.py:
```diff
@@ -88,51 +88,120 @@
+  unchanged_artifacts, modified_artifacts, _, deleted_artifacts = \
+      changes_between_snapshots(current_link.materials, current_link.products)
+  curr_step_materials = unchanged_artifacts.union(modified_artifacts,
+      deleted_artifacts)
```
Is there a difference between `set(current_link.materials)` and `curr_step_materials`? If not, I'd prefer just using the former instead of generating the latter via union over `modified_artifacts` and `deleted_artifacts`.

SolidifiedRay replied:
Yes, there is no difference.🤦 I have updated the changes.

lukpueh commented:
Apart from a tiny nit (see review above) this looks really good. Thanks for your persistence, @SolidifiedRay!

@IshaDave, I'd love if you could do some user testing here. Ideally while (re-)creating the layout for the task you are working on (https://github.com/in-toto/in-toto/issues/278). :)

lukpueh reviewed (approved):
🎉
